servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleProcessor.java

```diff
@@ -88,38 +67,34 @@ public void onError(final Throwable t) {
         terminate(TerminalNotification.error(t));
     }
 
-    private void terminate(@Nullable Object terminalSignal) {
-        if (terminalSignalUpdater.compareAndSet(this, TERMINAL_NULL, terminalSignal)) {
-            notifyListeners(terminalSignal);
-        }
-    }
-
-    private void notifyListeners(@Nullable Object terminalSignal) {
-        Subscriber<? super T> subscriber;
-        Throwable delayedCause = null;
+    private void terminateLateSubscriber(Subscriber<? super T> subscriber) {
+        Object terminalSignal = this.terminalSignal;
+        assert terminalSignal != TERMINAL_UNSET;
         if (terminalSignal instanceof TerminalNotification) {
             final Throwable error = ((TerminalNotification) terminalSignal).cause();
-            assert error != null : "Cause can't be null from TerminalNotification.error(..)";
-            while ((subscriber = subscribers.poll()) != null) {
-                try {
-                    subscriber.onError(error);
-                } catch (Throwable cause) {
-                    delayedCause = catchUnexpected(delayedCause, cause);
-                }
-            }
+            assert error != null;
+            subscriber.onError(error);
         } else {
             @SuppressWarnings("unchecked")
             final T value = (T) terminalSignal;
-            while ((subscriber = subscribers.poll()) != null) {
-                try {
-                    subscriber.onSuccess(value);
-                } catch (Throwable cause) {
-                    delayedCause = catchUnexpected(delayedCause, cause);
-                }
-            }
+            subscriber.onSuccess(value);
         }
-        if (delayedCause != null) {
-            throwException(delayedCause);
+    }
+
+    private void terminate(@Nullable Object terminalSignal) {
+        if (this.terminalSignal == TERMINAL_UNSET) {
+            // We must set terminalSignal before counter as we depend upon happens-before relationship for this value
+            // to be visible for any future late subscribers.
+            this.terminalSignal = terminalSignal;
+            if (terminalSignal instanceof TerminalNotification) {
+                final Throwable error = ((TerminalNotification) terminalSignal).cause();
```

Comment:
Instead of going back and forth between types, consider using `Consumer<Subscriber<...>>` type for `terminalSignal`. Then you can use `null` instead of `TERMINAL_UNSET` and simply pass the consumer into the `stack.close(...)`.

Reply:
interesting suggestion. this makes the code cleaner but may add allocation on the success path.

Reply:
`onSuccess` invokes this method and creates a lambda anyway, what will cause additional allocation? Assuming terminal methods are invoked only one time

Reply:
good point, you are right let me make this change!